src/idmap.rs

```diff
@@ -5,25 +5,124 @@
 //! Functionality for managing page tables with identity mapping.
 
 use crate::{
-    paging::{PhysicalAddress, Translation, VirtualAddress},
+    paging::{Attributes, MemoryRegion, PageTable, PhysicalAddress, Translation, VirtualAddress},
     MapError, Mapping,
 };
+use core::ptr::NonNull;
 
 /// Identity mapping, where every virtual address is either unmapped or mapped to the identical IPA.
 #[derive(Copy, Clone, Debug, Eq, PartialEq)]
 pub struct IdTranslation;
 
+impl IdTranslation {
+    fn virtual_to_physical(va: VirtualAddress) -> PhysicalAddress {
+        PhysicalAddress(va.0)
+    }
+}
+
 impl Translation for IdTranslation {
-    fn virtual_to_physical(&self, va: VirtualAddress) -> Result<PhysicalAddress, MapError> {
-        Ok(PhysicalAddress(va.0))
+    fn allocate_table(&self) -> (NonNull<PageTable>, PhysicalAddress) {
+        let table = PageTable::new();
```

Comment:
This confuses me. What kind of address does table.as_ptr() produce? Does this code require being executed via a 1:1 mapping?

Reply:
`as_ptr` returns a `*mut PageTable`, so it is a virtual address in the address space of the code being run. And yes, there is an assumption here that the code currently running is using the same 1:1 mapping as the page table it is manipulating, at least for the heap.

Reply:
Right. I think this severely limits the utility of the code, so maybe it is better to address this first before merging a partial solution? I don't know about Rialto, but OS kernel's generally use TTBR1, and so this code could be running from a non-1:1 address.

One potential solution (which could be part of a special implementation of the 'Translation' trait) is to rely on the 'AT' instruction, which performs a VA to PA translation using the currently active mapping. So instead of directly using the translation that is explicitly defined in the code, you derive the PA from the VA implicitly, and then convert that to a VA again (using software) if needed (note that there is no PA to VA translation instruction, for obvious reasons.)

But this is basically the reason I suggested avoiding going down this path if at all possible. Your brain is going to hurt :-)

Reply:
I expect anyone wanting to do anything beyond the very simple will need to write their own implementation of `Translation`. My main aim in this PR is to get `Translation` to a state where a range of different implementations are possible. I'm not quite sure how AT would help, but maybe we can discuss this more in person.

Reply:
I've changed the `physical_to_virtual` method to return a `NonNull<PageTable>` rather than a `VirtualAddress`, to clarify the distinction between the virtual addresses mapped by the pagetable, and the address space which the code managing the page table uses to access it, as they are not necessarily the same.